### PR TITLE
Add `EntityReference` type

### PR DESCRIPTION
**Pull request by foundrytom**
Part of #549.

### CODE REVIEW

**Comment by foundrytom** [comment]
@feltech lol, remember how we ran `pylint` locally? 🤯

File: src/openassetio-core/include/openassetio/EntityReference.hpp [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+/**
+ * EntityReference forms a strongly typed wrapper around arbitrary
+ * strings to ensure they have been validated by the target @ref manager
+ * before being used as an @ref entity_reference in the various entity
+ * related API calls.
+ *
+ * It can be assumed that if true is returned from @ref
+ * openassetio.hostApi.Manager.Manager.isEntityReference
+ * "isEntityReference" is True for any given string, then an
+ * EntityReference can be constructed from that string.
```
Comment (feltech) [inline src/openassetio-core/include/openassetio/EntityReference.hpp:21]: [fix] "if true is returned from .... is True for"

File: src/openassetio-core/include/openassetio/EntityReference.hpp [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+ *
+ * @warning EntityReferences should not be constructed directly by @ref
```
Comment (feltech) [inline src/openassetio-core/include/openassetio/EntityReference.hpp:23]: [fix] 
```suggestion
 * @warning EntityReferences should not be constructed directly by the @ref
```

File: src/openassetio-core/include/openassetio/EntityReference.hpp [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+ public:
+  /**
+   * Constructs an EntityReference around the supplied string.
+   */
+  explicit EntityReference(Str entityReferenceString)
+      : entityReferenceString_(std::move(entityReferenceString)) {}
+
+  ~EntityReference() = default;
+
+  [[nodiscard]] const Str& toString() const { return entityReferenceString_; }
```
Comment (feltech) [inline src/openassetio-core/include/openassetio/EntityReference.hpp:36]: [consider] Docstring

**Review by feltech — CHANGES_REQUESTED** [review]

File: src/openassetio-core/include/openassetio/EntityReference.hpp (hunk 1 above)
Comment (feltech) [reply src/openassetio-core/include/openassetio/EntityReference.hpp]: Tweaked in 68c3bf68

File: src/openassetio-core/include/openassetio/EntityReference.hpp (hunk 2 above)
Comment (feltech) [reply src/openassetio-core/include/openassetio/EntityReference.hpp]: Tweaked in 68c3bf68

File: src/openassetio-core/include/openassetio/EntityReference.hpp (hunk 3 above)
Comment (feltech) [reply src/openassetio-core/include/openassetio/EntityReference.hpp:39]: Added in 68c3bf68

File: src/openassetio-core/include/openassetio/EntityReference.hpp (hunk 1 above)
Comment (foundrytom) [reply src/openassetio-core/include/openassetio/EntityReference.hpp]: Doh! Thanks 🙄

**Pull request merged**